sdk/resourcemanager/Azure.ResourceManager/src/ArmClient.cs

```diff
@@ -59,6 +59,16 @@ public ArmClient(TokenCredential credential, string defaultSubscriptionId) : thi
         {
         }
 
+        /// <summary>
+        /// Initializes a new instance of the <see cref="ArmClient"/> class.
+        /// </summary>
+        /// <param name="credential"> A credential used to authenticate to an Azure Service. </param>
+        /// <param name="options"> The client parameters to use in these operations. </param>
+        /// <exception cref="ArgumentNullException"> If <see cref="TokenCredential"/> is null. </exception>
+        public ArmClient(TokenCredential credential, ArmClientOptions options) : this(credential, default, options)
```

Comment:
Lets remove this ctor and call one of the others that are provided.  I assume `ArmClient(TokenCredential credential, string defaultSubscriptionId, ArmClientOptions options)` will get used.  The `defaultSubscriptionId` can be left null, but if passed in it saves 1 API call when later using `armClient.GetDefaultSubscription()` as this value is cached for the lifetime of the client instance.

Reply:
Unfortunately, this constructor overload is required. Note that we are not calling the constructor ourselves, that is done by code inside Microsoft.Extensions.Azure. Here is that code: https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/extensions/Microsoft.Extensions.Azure/src/Internal/ClientFactory.cs#L22

Take this simplest example. We expect our `ArmClient` to be created using a `DefaultAzureCredential` and default options.
```csharp
var configuration = new ConfigurationBuilder()
    .AddInMemoryCollection()
    .Build();
var services = new ServiceCollection();
services
    .AddAzureClients(builder =>
        builder.AddArmClient(configuration));
var client = services
    .BuildServiceProvider()
    .GetRequiredService<ArmClient>();
```
With this overload existing, the client is created just fine. Without it, we get the following exception (details from Visual Studio from a local scratch project):
```
System.InvalidOperationException
  HResult=0x80131509
  Message=Unable to find matching constructor while trying to create an instance of ArmClient.
Expected one of the follow sets of configuration parameters:
1., defaultSubscriptionId

Found the following configuration keys:
  Source=Microsoft.Extensions.Azure
  StackTrace:
   at Microsoft.Extensions.Azure.ClientFactory.CreateClient(Type clientType, Type optionsType, Object options, IConfiguration configuration, TokenCredential credential)
   at Microsoft.Extensions.Azure.AzureClientFactoryBuilder.<>c__DisplayClass4_0`2.<Azure.Core.Extensions.IAzureClientFactoryBuilderWithConfiguration<Microsoft.Extensions.Configuration.IConfiguration>.RegisterClientFactory>b__0(TOptions options, TokenCredential credentials)
   at Microsoft.Extensions.Azure.ClientRegistration`1.GetClient(IServiceProvider serviceProvider, Object options, TokenCredential tokenCredential)
   at Microsoft.Extensions.Azure.AzureClientFactory`2.CreateClient(String name)
   at Microsoft.Extensions.Azure.AzureClientFactoryBuilder.<>c__15`2.<RegisterClientFactory>b__15_1(IServiceProvider provider)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.VisitRootCache(ServiceCallSite callSite, RuntimeResolverContext context)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteVisitor`2.VisitCallSite(ServiceCallSite callSite, TArgument argument)
   at Microsoft.Extensions.DependencyInjection.ServiceLookup.CallSiteRuntimeResolver.Resolve(ServiceCallSite callSite, ServiceProviderEngineScope scope)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.CreateServiceAccessor(Type serviceType)
   at System.Collections.Concurrent.ConcurrentDictionary`2.GetOrAdd(TKey key, Func`2 valueFactory)
   at Microsoft.Extensions.DependencyInjection.ServiceProvider.GetService(Type serviceType, ServiceProviderEngineScope serviceProviderEngineScope)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService(IServiceProvider provider, Type serviceType)
   at Microsoft.Extensions.DependencyInjection.ServiceProviderServiceExtensions.GetRequiredService[T](IServiceProvider provider)
   at Program.<<<Main>$>g__ResourceManagerTestaren|0_12>d.MoveNext() in C:\Users\Per.Stolpe\source\repos\ConsoleApp1\ConsoleApp1\Program.cs:line 687
   at Program.<<Main>$>d__0.MoveNext() in C:\Users\Per.Stolpe\source\repos\ConsoleApp1\ConsoleApp1\Program.cs:line 44
```

So, what is happening is that the internal `ClientFactory` from Microsoft.Extensions.Azure is looking for any constructor that has a parameter that takes an `ArmClientOptions`. Without this change, the only possible alternative is
```csharp
public ArmClient(TokenCredential credential, string defaultSubscriptionId, ArmClientOptions options)
```
The `credential` parameter is constructed using magic, but basically defaults into a `DefaultAzureCredential`, since it cannot find any relevant information about how it should construct it. This is just fine, and just as expected.

Now, it sees that we have no value for the parameter `defaultSubscriptionId`, and an exception is thrown. My overload is just this overload, but without the `defaultSubscriptionId` parameter. From eg.
```csharp
public ArmClient(TokenCredential credential) : this(credential, default, default)
```
we know that this value is not required.